src/steps/navigationSteps.ts

```diff
@@ -0,0 +1,121 @@
+  });
+
+  // if nav length is >1, then check the nav step prior as well
+  if (Object.keys(navigationSteps).length > 1) {
+    const prevNavStep = getStepsFromNavigation(currentNavIndex - 1);
+
+    Object.keys(prevNavStep).forEach((startMark: string) => {
+      const prevNavSteps = startMarkToStepsMap[startMark] ?? [];
+      Object.keys(prevNavSteps).forEach((step) => {
+        activeSteps[step] = true;
+      });
+    });
+  }
+
+  return activeSteps;
+}
+
+// start mark for navigationSteps data structure
+export const recordStartMark = (startMark: string) => {
+  const { navigationSteps } = steps;
+
+  // if we don't have any initial states recorded, initialize one
+  const navigationLength = Object.keys(navigationSteps).length > 0 ? Object.keys(navigationSteps).length : 1;
+  const lastNavIndex = navigationLength - 1;
+  const currentMarks = getStepsFromNavigation(lastNavIndex) || [];
+
+  const newCurrentMark = currentMarks;
+  newCurrentMark[startMark] = true;
+  navigationSteps[lastNavIndex] = newCurrentMark;
+}
+
+export const recordEndMark = (endMark: string) => {
+  const { navigationSteps, finalMarkToStepsMap } = steps;
+
+  const navigationLength = Object.keys(navigationSteps).length;
+  if (navigationLength === 0) {
+    return;
+  }
+
+  const currentNavIndex = navigationLength - 1;
+  const ongoingNavStep = getStepsFromNavigation(currentNavIndex);
+
+  if (!ongoingNavStep || !finalMarkToStepsMap[endMark]) {
+    return;
+  }
+
+  const endMarkMap = finalMarkToStepsMap[endMark];
+  if (!endMarkMap) {
+    return;
+  }
+
+  // loop through all steps where end mark is present and check if their start mark is present
+  // in the navigation state. If so, remove it to signal that the step has been completed
+  
+  Object.keys(endMarkMap).forEach((startMark) => {
+    if (ongoingNavStep[startMark]) {
+      const currentNavStep = getStepsFromNavigation(currentNavIndex) || {};
+
+      // Only remove the startMark, not other ongoing steps
+      currentNavStep[startMark] = false;
+      // if start mark is in navigation state, remove it to signal that the step has completed
+      navigationSteps[currentNavIndex] =  currentNavStep;
+    }
+
+    // if nav length is >1, then check the nav step prior as well
+    if (navigationLength > 1) {
+      const lastNavIndex = currentNavIndex - 1;
+      const lastNavStep = getStepsFromNavigation(lastNavIndex);
+      if (lastNavStep[startMark]) {
+        lastNavStep[startMark] = false;
+        navigationSteps[lastNavIndex] = lastNavStep;
+      }
+    }
+  });
+}
+
+// An API consumers must use to update internal state based on page/screen navigations.
+// Internal state must be updated when the following are true
+// 1. Journey marks are hit
+// 2. Page navigations occur
+// This function ensures the 2nd case is covered to help prevent "stale"
+// steps that don't hit their final mark because the users navigates
+// backwards in journey.
+export const incrementCujNavigation = () => {
```

Comment:
Avoid any mention of the word CUJ, instead let's use UJ as User Journey.
Across Perfume we just we will use two terms:
- User Journey
- User Journey Step